### PR TITLE
enhance metric support for product catalog service

### LINKED ISSUE #66: Review and enhance metric support for product catalog service (Go)
The following are requirements for metrics in the project. Not all services will meet all requirements.

- [ ] Service metrics should not be duplicative of trace-derived metrics.
- [ ] Services should emit multiple business metrics where appropriate.
- [ ] Services should emit trace exemplars for their metrics.
- [ ] Services should use push metrics where possible.
- [ ] Services should explicitly read baggage where needed.
- [ ] Collectors should perform metric transforms to normalize resource and service metrics.

Referencing issue: https://github.com/open-telemetry/opentelemetry-demo-webstore/issues/43

### PR DESCRIPTION
**Pull request by Frapschen**
Fixes #66 

## Changes

supprot runtime metrics for productcatalogservice.

For significant contributions please make sure you have completed the following items:

* [ ] Appropriate `CHANGELOG.md` updated for non-trivial changes
* [ ] Design discussion issue #66

### CODE REVIEW

**Comment by Frapschen** [comment]
@cartersocha, please review.

File: src/productcatalogservice/server.go [hunk 1]
```diff
@@ -76,6 +80,30 @@ func init() {
 	if err != nil {
 		log.Warnf("could not parse product catalog")
 	}
+
+	//init runtime metrics
```
Comment (cijothomas) [inline src/productcatalogservice/server.go:84]: is this using the prometheus client for Go? Should we not use the OTel library for Go instead?

Comment (fatsheep9146) [reply src/productcatalogservice/server.go:84]: Agreed, should use otel library.

**Comment by fatsheep9146** [comment]
> @cartersocha, please review. current implementation is not a push style, we really need it? it will add a new component,[a Prometheus Pushgateway](https://github.com/prometheus/pushgateway), I think it more costly, a static config for promethues is ok for me.

I think push gateway is not needed, we should use otel library to export metrics to otel-collector, the prometheus already scraped metrics from collector.

**Review by austinlparker — CHANGES_REQUESTED** [review]
We should be using the OTel Go Metrics library rather than Prometheus.

**Comment by Frapschen** [comment]
ok, I will use the OTel library to do it

**Comment by austinlparker** [comment]
Any updates on this?

**Comment by Frapschen** [comment]
I want wait this issue finish #2624, its a new version of instrumentation

**Comment by austinlparker** [comment]
> I want wait this issue finish [open-telemetry/opentelemetry-go-contrib#2624](https://github.com/open-telemetry/opentelemetry-go-contrib/issues/2624), its a new version of instrumentation

Does that issue have anyone actively working on it? It doesn't seem like it does... could we just use whatever the current version of go otel metrics instrumentation is, then when the new version comes out, update to that?

File: src/productcatalogservice/Dockerfile [hunk 2]
```diff
@@ -40,4 +40,5 @@ COPY ./src/productcatalogservice/products.json ./
 COPY --from=builder /go/bin/productcatalogservice/ ./
 
 EXPOSE ${PRODUCT_SERVICE_PORT}
+EXPOSE 2112
```
Comment (julianocosta89) [inline src/productcatalogservice/Dockerfile:43]: Can we add this port to the `.env` file, just to have more control over everything?

File: src/productcatalogservice/main.go [hunk 3]
```diff
@@ -16,53 +16,129 @@ package main
```
Comment (julianocosta89) [inline src/productcatalogservice/main.go:16]: I think you need to merge main into your branch.
This PR is undoing a couple of changes that we have merged already.

**Review by julianocosta89 — CHANGES_REQUESTED** [review]

**Comment by cartersocha** [comment]
Closing as the product catalog service was overhauled and will need to be updated.

**Pull request closed without merging**